### PR TITLE
Feature | logger initial implementation

### PR DESCRIPTION
**Pull request by AlexNDRmac**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -28,6 +28,14 @@ include(Bootstrap)
 include(Cppcheck)
 include(CodeCoverage)
 
+# Adds support for multithreading with the pthreads library
+# GNU gcc compiler must use explicit CXX flag for proper linker process
+FIND_PACKAGE(Threads REQUIRED)
+
+if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU")
```
Comment (sergeyklay) [inline CMakeLists.txt]:
```suggestion
if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
```

File: conanfile.txt [hunk 2]
```diff
@@ -8,11 +8,11 @@ gtest/1.10.0
...
 [options]
-spdlog:shared=True
 yaml-cpp:shared=False
 gtest:shared=False
 gtest:no_main=True
 
 [imports]
 bin, *.dll -> ./bin
 lib, *.dylib* -> ./bin
+lib, *.dylib* -> ./tests/bin
```
Comment (sergeyklay) [inline conanfile.txt]:
```suggestion
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -28,6 +28,14 @@ include(Bootstrap)
...
+  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
```
Comment (sergeyklay) [inline CMakeLists.txt]:
```suggestion
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
```

File: CMakeLists.txt [hunk 4]
```diff
@@ -28,6 +28,14 @@ include(Bootstrap)
 include(Cppcheck)
 include(CodeCoverage)
 
+# Adds support for multithreading with the pthreads library
+# GNU gcc compiler must use explicit CXX flag for proper linker process
+FIND_PACKAGE(Threads REQUIRED)
```
Comment (sergeyklay) [inline CMakeLists.txt]:
```suggestion
find_package(Threads REQUIRED)
```

File: CMakeLists.txt (hunk 3 above)
Comment (sergeyklay) [reply CMakeLists.txt]: Is it _required_ for whole project or just for a subset of project? Consider using `target_compile_options`. There is also `CMAKE_THREAD_LIBS_INIT`

Comment (AlexNDRmac) [reply CMakeLists.txt]: Yes. It'll be `required` for all project in future, when we'll use logger library

File: CMakeLists.txt [hunk 5]
```diff
@@ -28,6 +28,14 @@ include(Bootstrap)
...
+
+if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU")
+  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
+ENDIF()
```
Comment (sergeyklay) [inline CMakeLists.txt]:
```suggestion
endif()
```

File: CMakeLists.txt (hunk 3 above)
Comment (sergeyklay) [reply CMakeLists.txt]: I not think so. Right now it required only to link just logger library. We're not use pthread API

File: src/logger/logger.cpp [hunk 6]
```diff
@@ -0,0 +1,38 @@
+// This file is part of the Zephir.
+//
+// (c) Phalcon Team <team@zephir-lang.com>
+//
+// For the full copyright and license information, please view
+// the LICENSE file that was distributed with this source code.
+
+#include "logger.hpp"
+
+zephir::Logger::Logger() : _logger(std::make_unique<spdlog::logger>(this->_logger_name)) {}
```
Comment (sergeyklay) [inline src/logger/logger.cpp]:
```suggestion
zephir::Logger::Logger() : _logger(std::make_unique<spdlog::logger>(_logger_name)) {}
```

File: src/logger/logger.hpp [hunk 7]
```diff
@@ -8,24 +8,33 @@
...
+/// \brief A type definition for a unique pointer to a spdlog::logger instance.
+using LoggerPtr = std::unique_ptr<spdlog::logger>;
+
+/// \brief Wrap third party log library
+/// and provide minimalistic interface for Zephir log writer
+class Logger {
  private:
-  std::shared_ptr<spdlog::logger> _logger;
+  /// spdlog's global registry name
+  const std::string _logger_name = "console_logger";
```
Comment (sergeyklay) [inline src/logger/logger.hpp]:
```suggestion
  const std::string _channel = "console";
```

File: src/Logger/Logger.cpp [hunk 8]
```diff
@@ -7,7 +7,9 @@
 
 #include "Logger.hpp"
 
-zephir::Logger::Logger() : _logger(std::make_unique<spdlog::logger>(_channel)) {}
+zephir::Logger::Logger()
+    : _logger(std::make_unique<spdlog::logger>(
```
Comment (sergeyklay) [inline src/Logger/Logger.cpp]: Should we disable copy and move ctors then?

File: src/Logger/Logger.cpp [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  _logger->error(message);
+}
+
+void zephir::Logger::debug(const std::string& message) {
+  _logger->debug(message);
+}
+
+void zephir::Logger::exception(const std::exception& error,
+                               const Context& ctx) {
+  ///
```
Comment (sergeyklay) [inline src/Logger/Logger.cpp]: Could add an initial implementation with this?

File: src/Logger/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+# the LICENSE file that was distributed with this source code.
+
+# Add definition for Logger facade library.
+add_library(logger SHARED Logger.cpp)
+
+target_include_directories(
+  logger PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+                $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+
+target_link_libraries(logger PRIVATE CONAN_PKG::spdlog)
```
Comment (sergeyklay) [inline src/Logger/CMakeLists.txt]:
```suggestion
target_link_libraries(logger PRIVATE zephir_warnings CONAN_PKG::spdlog)
```

File: tests/Logger/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+file(GLOB LOGGER_TESTS ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)
+add_executable(LoggerTest ${ZEPHIR_TESTS_SOURCE_DIR}/main.cpp ${LOGGER_TESTS})
+
+target_include_directories(
+  LoggerTest
+  PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+  PRIVATE $<BUILD_INTERFACE:${ZEPHIR_TESTS_SOURCE_DIR}>
+          $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src>)
+
+target_link_libraries(LoggerTest PRIVATE logger CONAN_PKG::spdlog
```
Comment (sergeyklay) [inline tests/Logger/CMakeLists.txt]:
```suggestion
target_link_libraries(LoggerTest PRIVATE zephir_warnings logger CONAN_PKG::spdlog
```

File: tests/Logger/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+                                         CONAN_PKG::GTest)
```
Comment (sergeyklay) [inline tests/Logger/CMakeLists.txt]:
```suggestion
                                         CONAN_PKG::gtest)
```

**Comment by sergeyklay** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/zephir-lang/zephirc/pullRequest?prid=4643347&bid=15552997)

**Comment by sergeyklay** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/zephir-lang/zephirc/pullRequest?prid=4643347&bid=15552997)

**Comment by sergeyklay** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/zephir-lang/zephirc/pullRequest?prid=4643347&bid=15552997)

**Comment by AlexNDRmac** [comment]
@sergeyklay I think we can merge this state and continue working with logger ...

**Pull request merged**